index_documents.go

```diff
@@ -366,6 +370,169 @@ func (i Index) UpdateDocumentsInBatches(documentsPtr interface{}, batchSize int,
+	// into memory. However, this means that only part of the documents might
+	// be added successfully.
+
+	var (
+		responses []TaskInfo
+		header    []string
+		records   [][]string
+	)
+
+	sendCsvRecords := func(records [][]string) (*TaskInfo, error) {
+		b := new(bytes.Buffer)
+		w := csv.NewWriter(b)
+		w.UseCRLF = true // Keep output RFC 4180 compliant
+		err := w.WriteAll(records)
+		if err != nil {
+			return nil, fmt.Errorf("could not write CSV records: %w", err)
+		}
+
+		resp, err := i.UpdateDocumentsCsv(b.Bytes(), primaryKey...)
+		if err != nil {
+			return nil, err
+		}
+		return resp, nil
+	}
+
+	r := csv.NewReader(documents)
+	for {
+		// Read CSV record (empty lines and comments are already skipped by csv.Reader)
+		record, err := r.Read()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return nil, fmt.Errorf("could not write CSV record: %w", err)
+		}
+
+		// Store first record as header
+		if header == nil {
+			header = record
+			continue
+		}
+
+		// Add header record to every batch
+		if len(records) == 0 {
+			records = append(records, header)
+		}
+
+		records = append(records, record)
+
+		// After reaching batchSize (not counting the header record) assemble a CSV file and send records
+		if len(records) == batchSize+1 {
+			resp, err := sendCsvRecords(records)
+			if err != nil {
+				return nil, err
+			}
+			responses = append(responses, *resp)
+			records = nil
+		}
+	}
+
+	// Send remaining records as the last batch if there is any
+	if len(records) > 0 {
+		resp, err := sendCsvRecords(records)
+		if err != nil {
+			return nil, err
+		}
+		responses = append(responses, *resp)
+	}
+
+	return responses, nil
+}
+
+func (i Index) UpdateDocumentsNdjson(documents []byte, primaryKey ...string) (resp *TaskInfo, err error) {
+	return i.updateDocuments(documents, contentTypeNDJSON, primaryKey...)
+}
+
+func (i Index) UpdateDocumentsNdjsonInBatches(documents []byte, batchsize int, primaryKey ...string) (resp []TaskInfo, err error) {
+	return i.updateDocumentsNdjsonFromReaderInBatches(bytes.NewReader(documents), batchsize, primaryKey...)
+}
+
+func (i Index) updateDocumentsNdjsonFromReaderInBatches(documents io.Reader, batchSize int, primaryKey ...string) (resp []TaskInfo, err error) {
+	// NDJSON files supposed to contain a valid JSON document in each line, so
+	// it's safe to split by lines.
+	// Lines are read and sent continuously to avoid reading all content into
+	// memory. However, this means that only part of the documents might be
+	// added successfully.
+
+	sendNdjsonLines := func(lines []string) (*TaskInfo, error) {
+		b := new(bytes.Buffer)
+		for _, line := range lines {
+			_, err := b.WriteString(line)
+			if err != nil {
+				return nil, fmt.Errorf("Could not write NDJSON line: %w", err)
+			}
+			err = b.WriteByte('\n')
+			if err != nil {
+				return nil, fmt.Errorf("Could not write NDJSON line: %w", err)
+			}
+		}
+
+		resp, err := i.UpdateDocumentsNdjson(b.Bytes(), primaryKey...)
```

Comment:
Same for this method maybe a trick exist to avoid the duplication of this piece of code

Reply:
check it out, if you approve this approach, I'll use the same for other methods too where this pattern is occurring.

Reply:
It's better, but the problem is that almost all the method is duplicated from line 408 to 452 is possible to group all the duplicated code?

Reply:
How about using this same approach to implement a template design pattern?

Reply:
Check it out. Also, should I be doing the same for from reader functions?